system-contracts/test/L1Messenger.spec.ts

```diff
@@ -0,0 +1,526 @@
+import { ethers, network } from "hardhat";
+import type { L1Messenger } from "../typechain";
+import { L1MessengerFactory } from "../typechain";
+import { prepareEnvironment, setResult } from "./shared/mocks";
+import type { StateDiff } from "./shared/utils";
+import { compressStateDiffs, deployContractOnAddress, encodeStateDiffs, getCode, getWallets } from "./shared/utils";
+import { utils } from "zksync-web3";
+import type { Wallet } from "zksync-web3";
+import {
+  TEST_KNOWN_CODE_STORAGE_CONTRACT_ADDRESS,
+  TEST_L1_MESSENGER_SYSTEM_CONTRACT_ADDRESS,
+  TEST_BOOTLOADER_FORMAL_ADDRESS,
+  TWO_IN_256,
+} from "./shared/constants";
+import { expect } from "chai";
+import { BigNumber } from "ethers";
+import { randomBytes } from "crypto";
+
+describe("L1Messenger tests", () => {
+  let l1Messenger: L1Messenger;
+  let wallet: Wallet;
+  let l1MessengerAccount: ethers.Signer;
+  let knownCodeStorageAccount: ethers.Signer;
+  let bootloaderAccount: ethers.Signer;
+  let numberOfLogs: number = 0;
+  let numberOfMessages: number = 0;
+  let numberOfBytecodes: number = 0;
+  let stateDiffs: StateDiff[];
+  let encodedStateDiffs: string;
+  let compressedStateDiffs: string;
+  let numberOfStateDiffs: number;
+  let enumerationIndexSize: number;
+  let isService: boolean;
+  let key: Buffer;
+  let value: Buffer;
+  let message: Buffer;
+  let messageHash: string;
+  let txNumberInBlock: number;
+  let callResult: unknown;
+  let firstLog: string;
+  let secondLog: string;
+  let currentMessageLength: number;
+  let senderAddress: string;
+  let bytecode: string;
+  let lengthOfBytecode: number;
+  let lengthOfBytecodeBytes: string;
+  let bytecodeHash: string;
+  let currentMessageLengthBytes: string;
+  let version: string;
+  let enumerationIndexSizeBytes: string;
+  let stateDiffHash: string;
+  let verifyCompressedStateDiffsResult: unknown;
+  let numberOfStateDiffsBytes: string;
+  let compressedStateDiffsBuffer: Uint8Array;
+  let compressedStateDiffsLength: number;
+  let compressedStateDiffsSizeBytes: string;
+
+  before(async () => {
```

Comment:
this method looks a bit too large. Maybe it is possible to move out some functionality to separate functions?